File: frontend/src/services/package.js
/**
 * 套餐服务
 * 处理套餐类型相关的业务逻辑
 */

export class PackageService {
  // 套餐类别
  static PACKAGE_CATEGORIES = {
    count_based: '记次套餐',
    time_based: '时长套餐'
  }

  // 时长套餐类型
  static DURATION_TYPES = {
    monthly: '月卡',
    quarterly: '季卡',
    yearly: '年卡',
    custom: '自定义'
  }

  /**
   * 计算套餐到期时间
   * @param {string} durationType 时长类型
   * @param {Date} startDate 开始时间
   * @param {number} customDays 自定义天数
   * @returns {Date|null} 到期时间
   */
  static calculatePackageEndDate(durationType, startDate = new Date(), customDays = null) {
    if (!startDate) {
      startDate = new Date()
    }

    const endDate = new Date(startDate)

    switch (durationType) {
      case 'monthly':
        endDate.setMonth(endDate.getMonth() + 1)
        break
      case 'quarterly':
        endDate.setMonth(endDate.getMonth() + 3)
        break
      case 'yearly':
        endDate.setFullYear(endDate.getFullYear() + 1)
        break
      case 'custom':
        if (customDays && customDays > 0) {
          endDate.setDate(endDate.getDate() + customDays)
        } else {
          return null
        }
        break
      default:
        return null
    }

    return endDate
  }

  /**
   * 获取套餐状态文本
   * @param {Object} packageData 套餐数据
   * @returns {string} 状态文本
   */
  static getPackageStatusText(packageData) {
    const packageCategory = packageData.package_category || 'count_based'

    if (packageCategory === 'count_based') {
      // 记次套餐
      const remaining = packageData.remaining_lessons || 0
      const total = packageData.total_lessons || 0
      return `${remaining}/${total}次`
    } else {
      // 时长套餐
      if (packageData.unlimited_access) {
        return '永久有效'
      }

      const packageEndDate = packageData.package_end_date
      if (!packageEndDate) {
        return '永久有效'
      }

      // 处理字符串格式的日期
      let endDate
      if (typeof packageEndDate === 'string') {
        endDate = new Date(packageEndDate)
      } else {
        endDate = packageEndDate
      }

      if (isNaN(endDate.getTime())) {
        return '无效日期'
      }

      const now = new Date()
      const daysLeft = Math.ceil((endDate - now) / (1000 * 60 * 60 * 24))

      if (daysLeft <= 0) {
        return '已过期'
      } else if (daysLeft <= 30) {
        return `${daysLeft}天后过期`
      } else {
        return endDate.toLocaleDateString('zh-CN', {
          year: 'numeric',
          month: '2-digit',
          day: '2-digit'
        }) + '到期'
      }
    }
  }

  /**
   * 检查套餐是否有效
   * @param {Object} packageData 套餐数据
   * @returns {boolean} 是否有效
   */
  static isPackageValid(packageData) {
    const packageCategory = packageData.package_category || 'count_based'

    if (packageCategory === 'count_based') {
      // 记次套餐：检查剩余课程
      const remaining = packageData.remaining_lessons || 0
      return remaining > 0
    } else {
      // 时长套餐：检查到期时间
      if (packageData.unlimited_access) {
        return true
      }

      const packageEndDate = packageData.package_end_date
      if (!packageEndDate) {
        return true
      }

      // 处理字符串格式的日期
      let endDate
      if (typeof packageEndDate === 'string') {
        endDate = new Date(packageEndDate)
      } else {
        endDate = packageEndDate
      }

      if (isNaN(endDate.getTime())) {
        return false
      }

      return new Date() <= endDate
    }
  }

  /**
   * 处理创建学生时的套餐数据
   * @param {Object} createData 创建数据
   * @returns {Object} 处理后的数据
   */
  static processPackageDataForCreation(createData) {
    const packageCategory = createData.package_category || 'count_based'

    // 设置兼容字段
    const processedData = { ...createData }
    processedData.package_type = createData.original_package_type || '1v1'

    if (packageCategory === 'count_based') {
      // 记次套餐：设置初始剩余课程数
      const totalLessons = createData.total_lessons || 0
      processedData.remaining_lessons = totalLessons
      // 清空时长套餐相关字段
      processedData.package_end_date = null
      processedData.package_duration_type = null
      processedData.package_duration_days = null
      processedData.unlimited_access = false
    } else {
      // 时长套餐：计算到期时间
      const durationType = createData.package_duration_type
      const customDays = createData.package_duration_days
      const unlimited = createData.unlimited_access || false

      if (!unlimited && durationType) {
        const endDate = this.calculatePackageEndDate(durationType, new Date(), customDays)
        processedData.package_end_date = endDate
      } else {
        processedData.package_end_date = null
      }

      // 清空记次套餐相关字段
      processedData.total_lessons = null
      processedData.remaining_lessons = null
    }

    return processedData
  }

  /**
   * 获取套餐显示信息
   * @param {Object} packageData 套餐数据
   * @returns {Object} 显示信息
   */
  static getPackageInfoDisplay(packageData) {
    const packageCategory = packageData.package_category || 'count_based'
    const originalPackageType = packageData.original_package_type || '1v1'

    const result = {
      category: this.PACKAGE_CATEGORIES[packageCategory] || packageCategory,
      type: originalPackageType === '1v1' ? '一对一' : '一对多',
      status: this.getPackageStatusText(packageData),
      valid: this.isPackageValid(packageData) ? '有效' : '无效'
    }

    if (packageCategory === 'time_based') {
      const durationType = packageData.package_duration_type
      if (durationType) {
        result.duration = this.DURATION_TYPES[durationType] || durationType
      }

      if (packageData.unlimited_access) {
        result.duration = '永久'
      }
    }

    return result
  }

  /**
   * 验证套餐数据
   * @param {Object} packageData 套餐数据
   * @returns {Array} 错误信息数组
   */
  static validatePackageData(packageData) {
    const errors = []
    const packageCategory = packageData.package_category || 'count_based'

    if (packageCategory === 'count_based') {
      const totalLessons = packageData.total_lessons
      if (!totalLessons || totalLessons <= 0) {
        errors.push('记次套餐必须设置总课程数')
      }
    } else {
      const durationType = packageData.package_duration_type
      const unlimited = packageData.unlimited_access

      if (!unlimited && !durationType) {
        errors.push('时长套餐必须选择时长类型或设置永久有效')
      }

      if (durationType === 'custom') {
        const customDays = packageData.package_duration_days
        if (!customDays || customDays <= 0) {
          errors.push('自定义时长套餐必须设置有效天数')
        }
      }
    }

    return errors
  }
}

File: backend/api_server/services/package.py
"""
套餐服务模块
处理套餐类型相关的业务逻辑，包括记次套餐和时长套餐的管理
"""

from datetime import datetime, timedelta
from typing import Optional, Dict, Any
from dateutil.relativedelta import relativedelta


class PackageService:
    """套餐服务类"""

    # 套餐类别定义
    PACKAGE_CATEGORIES = {
        "count_based": "记次套餐",
        "time_based": "时长套餐"
    }

    # 时长套餐类型定义
    DURATION_TYPES = {
        "monthly": "月卡",
        "quarterly": "季卡",
        "yearly": "年卡",
        "custom": "自定义"
    }

    @staticmethod
    def calculate_package_end_date(duration_type: str, start_date: datetime = None,
                                 custom_days: int = None) -> Optional[datetime]:
        """
        计算套餐到期时间

        Args:
            duration_type: 时长类型 (monthly, quarterly, yearly, custom)
            start_date: 开始时间，默认为当前时间
            custom_days: 自定义天数(仅custom类型使用)

        Returns:
            到期时间，如果没有设置则返回None
        """
        if not start_date:
            start_date = datetime.now()

        if duration_type == "monthly":
            return start_date + relativedelta(months=1)
        elif duration_type == "quarterly":
            return start_date + relativedelta(months=3)
        elif duration_type == "yearly":
            return start_date + relativedelta(years=1)
        elif duration_type == "custom" and custom_days:
            return start_date + timedelta(days=custom_days)

        return None

    @staticmethod
    def get_package_status_text(package_data: Dict[str, Any]) -> str:
        """
        获取套餐状态文本

        Args:
            package_data: 包含套餐信息的字典

        Returns:
            套餐状态文本
        """
        package_category = package_data.get("package_category", "count_based")

        if package_category == "count_based":
            # 记次套餐
            remaining = package_data.get("remaining_lessons", 0)
            total = package_data.get("total_lessons", 0)
            return f"{remaining}/{total}次"
        else:
            # 时长套餐
            if package_data.get("unlimited_access", False):
                return "永久有效"

            package_end_date = package_data.get("package_end_date")
            if not package_end_date:
                return "永久有效"

            # 处理字符串格式的日期
            if isinstance(package_end_date, str):
                try:
                    package_end_date = datetime.fromisoformat(package_end_date.replace('Z', '+00:00'))
                except:
                    return "无效日期"

            days_left = (package_end_date - datetime.now()).days
            if days_left <= 0:
                return "已过期"
            elif days_left <= 30:
                return f"{days_left}天后过期"
            else:
                return package_end_date.strftime("%Y/%m/%d到期")

    @staticmethod
    def is_package_valid(package_data: Dict[str, Any]) -> bool:
        """
        检查套餐是否有效

        Args:
            package_data: 包含套餐信息的字典

        Returns:
            套餐是否有效
        """
        package_category = package_data.get("package_category", "count_based")

        if package_category == "count_based":
            # 记次套餐：检查剩余课程
            remaining = package_data.get("remaining_lessons", 0)
            return remaining > 0
        else:
            # 时长套餐：检查到期时间
            if package_data.get("unlimited_access", False):
                return True

            package_end_date = package_data.get("package_end_date")
            if not package_end_date:
                return True

            # 处理字符串格式的日期
            if isinstance(package_end_date, str):
                try:
                    package_end_date = datetime.fromisoformat(package_end_date.replace('Z', '+00:00'))
                except:
                    return False

            return datetime.now() <= package_end_date

    @staticmethod
    def process_package_data_for_creation(create_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        处理创建学生时的套餐数据

        Args:
            create_data: 创建学生的原始数据

        Returns:
            处理后的套餐数据
        """
        package_category = create_data.get("package_category", "count_based")

        # 设置兼容字段
        processed_data = create_data.copy()
        processed_data["package_type"] = create_data.get("original_package_type", "1v1")

        if package_category == "count_based":
            # 记次套餐：设置初始剩余课程数
            total_lessons = create_data.get("total_lessons", 0)
            processed_data["remaining_lessons"] = total_lessons
            # 清空时长套餐相关字段
            processed_data["package_end_date"] = None
            processed_data["package_duration_type"] = None
            processed_data["package_duration_days"] = None
            processed_data["unlimited_access"] = False
        else:
            # 时长套餐：计算到期时间
            duration_type = create_data.get("package_duration_type")
            custom_days = create_data.get("package_duration_days")
            unlimited = create_data.get("unlimited_access", False)

            if not unlimited and duration_type:
                end_date = PackageService.calculate_package_end_date(
                    duration_type, custom_days=custom_days
                )
                processed_data["package_end_date"] = end_date
            else:
                processed_data["package_end_date"] = None

            # 清空记次套餐相关字段
            processed_data["total_lessons"] = None
            processed_data["remaining_lessons"] = None

        return processed_data

    @staticmethod
    def get_package_info_display(package_data: Dict[str, Any]) -> Dict[str, str]:
        """
        获取套餐显示信息

        Args:
            package_data: 包含套餐信息的字典

        Returns:
            包含显示信息的字典
        """
        package_category = package_data.get("package_category", "count_based")
        original_package_type = package_data.get("original_package_type", "1v1")

        result = {
            "category": PackageService.PACKAGE_CATEGORIES.get(package_category, package_category),
            "type": original_package_type,
            "status": PackageService.get_package_status_text(package_data),
            "valid": "有效" if PackageService.is_package_valid(package_data) else "无效"
        }

        if package_category == "time_based":
            duration_type = package_data.get("package_duration_type")
            if duration_type:
                result["duration"] = PackageService.DURATION_TYPES.get(duration_type, duration_type)

            if package_data.get("unlimited_access", False):
                result["duration"] = "永久"

        return result